scalanet/src/io/iohk/scalanet/peergroup/kademlia/KNetwork.scala

```diff
@@ -76,13 +105,26 @@ object KNetwork {
       maybeError.fold(Task.unit)(_ => channel.close())
     }
 
-    private def nodesTask(
+    private def sendResponse(
         channel: Channel[A, KMessage[A]]
-    ): Nodes[A] => Task[Unit] = { nodes =>
+    ): KMessage[A] => Task[Unit] = { message =>
       channel
-        .sendMessage(nodes)
+        .sendMessage(message)
         .timeout(requestTimeout)
         .doOnFinish(_ => channel.close())
     }
+
+    private def serverTemplate[Request <: KRequest[A], Response <: KResponse[A]](
+        pf: PartialFunction[KMessage[A], Request]
+    ): Observable[(Request, KMessage[A] => Task[Unit])] = {
+      peerGroup.server().collectChannelCreated.mapTask { channel: Channel[A, KMessage[A]] =>
+        channel.in
+          .collect(pf)
+          .map(request => (request, sendResponse(channel)))
+          .headL
+          .timeout(requestTimeout)
+          .doOnFinish(closeIfAnError(channel))
```

Comment:
from `timeout` documentation
```
Returns a Task that mirrors the source Task but that triggers a TimeoutException in case the given duration passes without the task emitting any item.
```
I agree about the race condition issue if the same message is "accepted" by `pf` after `collect`.

I believe that if the underlying channel does not deliver a message that passes the filter provided by `collect(pf)` before `requestTimeout` elapses, the channel will be closed and requests will be lost. This can happen if the underlying peergrou is TCP for example, where channels are received when connections start but probably would not happen (if the `requestTimeout` is big enough) in UDP where channels are created when the first message arrive.